src/overlords/offense/controllerAttacker.ts

```diff
@@ -0,0 +1,106 @@
+// Controller attacker overlord.  Spawn CLAIM creeps to mass up on a controller and attack all at once
+
+import {Zerg} from '../../zerg/Zerg';
+import {OverlordPriority} from '../../priorities/priorities_overlords';
+import {Overlord} from '../Overlord';
+import {DirectiveControllerAttack} from '../../directives/offense/controllerAttack';
+import {CreepSetup} from '../CreepSetup';
+import {profile} from '../../profiler/decorator';
+import {log} from '../../console/log';
+import {SpawnGroup} from 'logistics/SpawnGroup';
+
+const ControllerAttacker = new CreepSetup('ctrlAttack', {
+	pattern: [CLAIM, MOVE],
+	sizeLimit: Infinity,
+	ordered: true
+});
+
+interface ControllerAttackerMemory extends CreepMemory {
+	attackPos: protoPos;
+}
+
+@profile
+export class ControllerAttackerOverlord extends Overlord {
+
+	attackers: ControllerAttackerZerg[];
+	scouts: Zerg[];
+	attackPos: RoomPosition[] | undefined;
+	spawns: StructureSpawn[] | undefined;
+	readyAtTick: number;
+
+	currentThreshold: number;
+
+	settings: {
+		baseThreshold: number,
+		ticksTilThresholdDecrease: number,
+	};
+
+	constructor(directive: DirectiveControllerAttack, priority = OverlordPriority.offense.controllerAttack) {
+		super(directive, 'controllerAttack', priority);
+
+		this.attackers = _.map(this.creeps(ControllerAttacker.role), creep => new ControllerAttackerZerg(creep));
+		this.attackPos = directive.attackPos;
+		this.readyAtTick = directive.readyAtTick;
+		if (directive.roomName) {
+			this.spawnGroup = new SpawnGroup(directive.roomName, { requiredRCL: 3 });
+		}
+	}
+
+	init() {
+		// TODO: Prespawn attackers to arrive as cooldown disappears
+		if (this.attackPos && Game.time >= this.readyAtTick) {
+			this.wishlist(this.attackPos.length, ControllerAttacker);
+			// hack to get the SpawnGroup to initialize until it's fixed in Overmind.ts
+			if (this.spawnGroup) {
+				this.spawnGroup.init();
+			}
+		}
+	}
+
+	run() {
+		if (this.attackers) {
+			for (let attacker of this.attackers) {
+				if (!attacker.memory.attackPos) {
+					attacker.memory.attackPos = _.filter(this.attackPos!, pos => !_.find(this.attackers, otherAttacker => otherAttacker.memory.attackPos ? derefRoomPosition(otherAttacker.memory.attackPos).isEqualTo(pos) : false))[0];
```

Comment:
this is a little confusing. a better way to do this would be similar to how `BunkerQueenOverlord` works: make the attackers be sorted by id and just assign each attacker the corresponding attack position